### PR TITLE
docs: Update type annotation and docstring content and formatting

### PR DESCRIPTION
**Pull request by andriymujin**

### CODE REVIEW

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 1]
```diff
@@ -16,20 +16,19 @@
...
 # logging
-from logging import getLogger
-log = getLogger(__name__)
+import logging
+log = logging.getLogger(__name__)
 
-class VisionControllerClient(object):
-    """Mujin Vision Controller client for binpicking tasks.
-    """
+class VisionClient(object):
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: Can we separate the name change into another MR like https://github.com/mujin/mujinvisioncontrollerclientpy/pull/39 ?

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 2]
```diff
@@ -257,15 +281,16 @@ def StopTask(self, taskId=None, taskIds=None, taskType=None, taskTypes=None, cyc
...
+            fireandforget (bool, optional): If True, does not wait for the command to finish and returns immediately. The command remains queued on the server. (Default: False)
             timeout (float, optional): Time in seconds after which the command is assumed to have failed. (Default: 2.0)
 
         Returns:
             dict: A dictionary with the structure:
 
-            isStopped (bool): true, if the specific taskId or set of tasks with a specific taskType(s) is stopped
+                - isStopped (bool): true, if the specific taskId or set of tasks with a specific taskType(s) is stopped
         """
+        log.verbose('Stopping detection thread...')
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: Can you put these additional logs into a separate MR as well? It's not clear that they won't be spammy.

**Review by felixvd — COMMENTED** [review]
Just move the name change and the logging into separate MR and this should be good. Thanks!

**Comment by andriymujin** [comment]
@felixvd I removed the changes that you've mentioned. I will submit a PR for them after this one merges.

PS: I couldnt find a good way to create a chain of Pull requests. Creating a Pull request with logging into this branch is not what we want.

**Comment by andriymujin** [comment]
@florisatmujin , since Felix is OOO, please take a look if those changes are ok to merge.

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 3]
```diff
@@ -167,12 +171,22 @@ def _WaitForResponse(self, recvjson=True, timeout=None, command=None):
...
         """Returns whether the client is waiting for response on the command socket, and caller should call WaitForResponse().
         """
         return self._commandsocket.IsWaitingReply()
 
     def _SendConfiguration(self, configuration, fireandforget=False, timeout=2.0, checkpreempt=True, recvjson=True):
         # type: (Dict, bool, float, bool, bool) -> Optional[Dict]
+        """Sends a configuration command.
+
+        Args:
+            configuration (dict):
```
Comment (florisatmujin) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: Can we add a description here?

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 4]
```diff
@@ -185,16 +199,21 @@ def _SendConfiguration(self, configuration, fireandforget=False, timeout=2.0, ch
     #
 
     def StartObjectDetectionTask(self, taskId=None, systemState=None, visionTaskParameters=None, timeout=2.0, **ignoredArgs):
-        # type: (Optional[types.SystemState], Optional[types.visionTaskObjectDetectionParametersSchema], float, Dict) -> Optional[Dict]
+        # type: (Optional[str], Optional[types.SystemState], Optional[types.visionTaskObjectDetectionParametersSchema], float, Dict) -> Optional[Dict[str, str]]
```
Comment (florisatmujin) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: **ignoredArgs -> is that a dict or Optional[Dict] ?

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 5]
```diff
@@ -38,38 +38,41 @@ class VisionControllerClient(object):
     _checkpreemptfn = None # called periodically when in a loop
     
     _subscriber = None # an instance of ZmqSubscriber, used for subscribing to the state
-    
+
+    _deprecated = None # used to mark arguments as deprecated (set argument default value to this)
+
     def __init__(self, hostname='127.0.0.1', commandport=7004, ctx=None, checkpreemptfn=None, reconnectionTimeout=40, callerid=None):
         # type: (str, int, Optional[zmq.Context], Optional[Callable], float, Optional[str]) -> None
```
Comment (florisatmujin) [inline mujinvisioncontrollerclient/visioncontrollerclient.py:44]: To make sure, when there is a default value like `reconnectionTimeout=40` then it is not `Optional`. That is only in the case when the value can be `None`

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 3 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: Done. 

PS: Also added a docstring to the ExecuteCommand so it is matches with SendConfiguration.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 4 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: Thank you for pointing it out! I was just following the existing code, but it wasnt a good approach.

The type here should actually be Any. See https://stackoverflow.com/questions/37031928/type-annotations-for-args-and-kwargs .

I've also checked it with Pylance. It discovered multiple other type issues. Fixed them now. PTAL.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 5 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py:44]: Yes.

It clashed a bit with how we set optional in docstrings. I dont like it myself. In the future I would like to get rid of type and optional annotation in docstrings at all actually. Since now it will become redundant. But it is better to do in a separate MR since it may be more controversial.

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 6]
```diff
@@ -107,9 +109,20 @@ def SetDestroy(self):
             self._commandsocket.SetDestroy()
         if self._configurationsocket is not None:
             self._configurationsocket.SetDestroy()
-    
+
     def _ExecuteCommand(self, command, fireandforget=False, timeout=2.0, recvjson=True, checkpreempt=True, blockwait=True):
-        # type: (Dict, bool, float, bool, bool, bool) -> Optional[Dict]
+        # type: (Dict, bool, float, bool, bool, bool) -> Any
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py:114]: Why is this Any? This will throw errors when the other functions that do `return self._ExecuteCommand` have a type hint returning Optional[Dict].

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 7]
```diff
@@ -167,12 +182,24 @@ def _WaitForResponse(self, recvjson=True, timeout=None, command=None):
         return self._ProcessResponse(response, command=command, recvjson=recvjson)
 
     def IsWaitingResponse(self):
+        # type: () -> bool
         """Returns whether the client is waiting for response on the command socket, and caller should call WaitForResponse().
         """
+        assert self._commandsocket is not None
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py:188]: Please do not change logic in this MR

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 8]
```diff
@@ -185,19 +212,26 @@ def _SendConfiguration(self, configuration, fireandforget=False, timeout=2.0, ch
...
+
+                A dictionary with the structure:
+
+                - taskId (str): The taskId of the created task
         """
         log.verbose('Starting detection thread...')
-        command = {'command': 'StartObjectDetectionTask'}  # type: Dict[str, Any]
+        command = {
+            'command': 'StartObjectDetectionTask',
+        }  # type: Dict[str, Any]
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py:234]: I think the type checker expects this to be on the first line

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 9]
```diff
@@ -207,19 +241,26 @@ def StartObjectDetectionTask(self, taskId=None, systemState=None, visionTaskPara
         return self._ExecuteCommand(command, timeout=timeout)
 
     def StartContainerDetectionTask(self, taskId=None, systemState=None, visionTaskParameters=None, timeout=2.0, **ignoredArgs):
-        # type: (Optional[types.SystemState], Optional[types.visionTaskContainerDetectionParametersSchema], float, Dict) -> Optional[Dict]
+        # type: (Optional[str], Optional[Union[types.SystemState, Dict]], Optional[Union[types.visionTaskContainerDetectionParametersSchema, Dict]], float, Any) -> Optional[Dict[str, str]]
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: Why Optional[Union[types.SystemState, Dict]]? I think the docstring was outdated here.

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 10]
```diff
@@ -207,19 +241,26 @@ def StartObjectDetectionTask(self, taskId=None, systemState=None, visionTaskPara
...
         """Starts container detection thread to continuously detect a container. the vision server will send detection results directly to mujin controller.
 
         Args:
-            taskId (str, optional): If specified, the specific taskId to use
-            systemState (types.SystemState or dict): The state of the system. Used to select the profile that the vision task will use. See "Profile Selection" documentation for more details.
-            visionTaskParameters (types.visionTaskContainerDetectionParametersSchema or dict): Parameters for the container detection task. These take precedence over the base profile selected via the system state, but are overwritten by the overwrite profile.
+            taskId (str, optional): If specified, the specific taskId to use.
+            systemState (types.SystemState | dict, optional): The state of the system. Used to select the profile that the vision task will use. See "Profile Selection" documentation for more details.
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: I don't think the syntax is understood when the doc is generated.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 9 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: Oh, I see. Fixed then.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 10 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: The docs with this syntax are generated the same way as without it. The type in parentheses is displayed as is in both cases.

While the parameter type is taken from the type comment. E.g.:
```
StartObjectDetectionTask(taskId: Optional[str] = None, systemState: Optional[Union[mujinvisiontypes.SystemState, Dict]] = None, visionTaskParameters: Optional[Union[mujinvisiontypes.visionTaskObjectDetectionParametersSchema, Dict]] = None, timeout: float = 2.0, **ignoredArgs: Any)→ Optional[Dict[str, str]]
```

But taking into account your comment above, it doesnt matter, lets remove the dict part.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 8 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py:234]: No, at least Pylance expects the comment here. It shows an error if I put the comment on the first line.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 6 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py:114]: It is Any, because depending on `recvjson` value it may return not dict. We could also be Union[Dict, str, None] . But its actually the opposite of what you say - with Any Pylance doesnt show any errors, while with Union[Dict, str, None] or even Optional[Dict] we would have to do bunch of type checks. E.g. for return value not being None before doing `SendAndReceive(...).get('parametervalue', None)` (example from another client).

Of course ideally we would like to not use Any unless we have to. But the way our code is written makes it quite hard for now. And _ExecuteCommand is private function - so users of the client wont interact with it anyway.

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 11]
```diff
@@ -16,12 +16,11 @@
 from . import ugettext as _
 
 # logging
-from logging import getLogger
-log = getLogger(__name__)
+import logging
+log = logging.getLogger(__name__)  # type: Any
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: Why add this type?

File: mujinvisioncontrollerclient/visioncontrollerclient.py [hunk 12]
```diff
@@ -157,22 +172,34 @@ def _WaitForResponse(self, recvjson=True, timeout=None, command=None):
             raise VisionControllerClientError(_('Waiting on command "%(commandName)s" when wait signal is not on') % {
                 'commandName': commandName,
             }, errortype='invalidwait')
-        
+
         try:
-            response = self._commandsocket.ReceiveCommand(timeout=timeout, recvjson=recvjson)
+            response = self._commandsocket.ReceiveCommand(timeout=timeout, recvjson=recvjson)  # type: ignore # timeout can be None, but there is no type annotations in ReceiveCommand to indicate it.
```
Comment (felixvd) [inline mujinvisioncontrollerclient/visioncontrollerclient.py]: I'd rather fix it [over there](https://github.com/mujin/mujinplanningclientpy/blob/master/python/mujinplanningclient/zmqclient.py#L400C30-L400C37) or leave it for later instead of adding this `type: ignore`.

**Review by felixvd — COMMENTED** [review]
LGTM except for the `type: ignore`

**Comment by felixvd** [comment]
Could you also run autotester since this Github repo isn't connected to our CI? Thanks

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 11 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: Otherwise Pylance complains:
```
Cannot access member "verbose" for type "Logger"
  Member "verbose" is unknown
```

This is for sure a hack. But I dont know how to do better. Overall, I dont know why it works on master as I cant find the definition of `verbose` myself.

If you know more about it I would appreciate the help with fixing it.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 12 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: Agree, good point. Lets leave it for late. Done.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 11 above)
Comment (andriymujin) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: Now that I think about it - it was a bad idea to hide this error as well. Will remove the type: Any regardless of whether there is another way to fix it.

**Comment by felixvd** [comment]
I haven't looked at this in detail, but I saw the same error and assumed
it's a Python2/3 issue. The system must be using a different logging
library than the type checker. Let's wait until we migrated to Python 3
before we spend time on this.

General comment: Error messages are very noisy with type checking on. I
also found myself making changes like this to silence warnings, but it's
not always productive. It might be a good idea to turn it off until we are
a bit further in the process, or you want to work on types specifically.

Also, if we use the `JSONSchema` plugin, we need to use mypy instead of
Pylance, and their errors are not the same.

File: mujinvisioncontrollerclient/visioncontrollerclient.py (hunk 11 above)
Comment (felixvd) [reply mujinvisioncontrollerclient/visioncontrollerclient.py]: The response I emailed was supposed to go into this thread

**Comment by andriymujin** [comment]
Test look ok (#607091,  #607179, #607317). There are some occasional failures on green tests. But no tests that would be falling across the runs. So seems like those are just flakes.

Merging
